diff -Naur zentrack_2-5-5-1/includes/ZenFieldMap.class zentrack_2-5-5-2/includes/ZenFieldMap.class
--- zentrack_2-5-5-1/includes/ZenFieldMap.class	2005-05-08 07:41:02.000000000 -0700
+++ zentrack_2-5-5-2/includes/ZenFieldMap.class	2005-05-15 16:44:30.000000000 -0700
@@ -167,7 +167,7 @@
     "custom_date"     => array(   true,   array(1,7)            , 'date'   ,      null, false,  false,  true),
     "custom_menu"     => array(   true,   array(1,3,6)          , 'string' ,      null, false,  false,  true),
     "custom_text"     => array(  false,   array(1,2)            , 'text'   ,      null, false,  false,  true),
-    "hours"           => array(  false,   array(2)              , 'int'    ,      null, false,  false,  true),
+    "hours"           => array(  false,   array(2)              , 'float'  ,      null, false,  false,  true),
     "comments"        => array(  false,   array(2)              , 'text'   ,      null, false,  false,  true)
   ),
   
@@ -519,17 +519,17 @@
   /**
    * Updates the database and the session with new field map values
    *
-   * @param array $newvals an array structured identical to getFieldMap() results
+   * @param array $updates an array structured identical to getFieldMap() results
    * @param array $newspacers array ( 'field_name' => array('field_label', 'is_visible', 'sort_order'), ... )
    * @param array $oldspacers array ( 'field_name'=>field_map_id, ... )
    * @return array(updates, attempts)
    */
-  function updateFieldMap( $view, $newvals, $newspacers = null, $oldspacers = null ) {
+  function updateFieldMap( $view, $updates, $newspacers = null, $oldspacers = null ) {
     $updateVals = array();
     $fullMap = $this->getFieldMap();
     foreach($fullMap[$view] as $k=>$v) {
-      if( $newvals[$k] && $newvals[$k] != $v ) {
-        $updateVals[] = $newvals[$k];
+      if( $updates[$k] && !Zen::arrayEquals($updates[$k], $v) ) {
+        $updateVals[] = $updates[$k];
       }
     }
     $i=0;
@@ -680,7 +680,8 @@
     $this->_zen->addDebug('_getChoices', "Generated ".count($vals)." choices for {$view}->{$field_name}", 3);
     
     if( $view == 'search_form' && $field['num_rows'] == 1 ) {
-      $vals = array(''=>'-any-') + $vals ;
+      $vals = Zen::mergeArrays( array('' => '-any-'), $vals );
+      //$vals = array_merge(array(''=>'-any-'),$vals);
     }
     if( !isset($this->_fieldVals[$view]) ) { $this->_fieldVals[$view] = array(); }
     $this->_fieldVals[$view][$field_name] = $vals;
diff -Naur zentrack_2-5-5-1/includes/ZenSessionManager.class zentrack_2-5-5-2/includes/ZenSessionManager.class
--- zentrack_2-5-5-1/includes/ZenSessionManager.class	2005-05-01 23:59:35.000000000 -0700
+++ zentrack_2-5-5-2/includes/ZenSessionManager.class	2005-05-15 16:44:30.000000000 -0700
@@ -104,7 +104,7 @@
   function clearDataType( $type ) {
     if( isset($_SESSION) ) {
       if( isset($_SESSION['ztDataTypes']) ) {
-        $this->_zen->addDebug("clearDataType", "$type: cleared");
+        $this->_zen->addDebug("clearDataType", "$type: cleared",3);
         unset($_SESSION['ztDataTypes'][$type]);
       }
     }
diff -Naur zentrack_2-5-5-1/includes/db.class zentrack_2-5-5-2/includes/db.class
--- zentrack_2-5-5-1/includes/db.class	2005-05-08 07:41:02.000000000 -0700
+++ zentrack_2-5-5-2/includes/db.class	2005-05-15 16:44:30.000000000 -0700
@@ -69,7 +69,7 @@
     // returns an array of results
     // from $query
     $this->switchQueryMode(0);
-    $recordSet = &$this->db_link->Execute($query);
+    $recordSet = $this->db_link->Execute($query);
     if( $recordSet ) {
       $vars = $recordSet->getArray();
       if( is_array($vars) && count($vars) > 0 )
@@ -165,7 +165,7 @@
     $query = "INSERT INTO $table ($cols) VALUES($vals)";
     $res = $this->db_link->Execute($query);
     $this->addDebug("db_insert","[$res]".$query,3);
-    if( !$res && strpos($this->database_type,"mssql")===0 ) {
+    if( !$res && strlen(strpos($this->database_type,"mssql")) ) {
       // deal with sql server id issues by automagically
       // incrementing the counter when a problem occurs
       $query = "SELECT max($id_field) FROM $table";
diff -Naur zentrack_2-5-5-1/includes/editAgreementSubmit.php zentrack_2-5-5-2/includes/editAgreementSubmit.php
--- zentrack_2-5-5-1/includes/editAgreementSubmit.php	2005-05-01 23:59:35.000000000 -0700
+++ zentrack_2-5-5-2/includes/editAgreementSubmit.php	2005-05-15 16:44:30.000000000 -0700
@@ -4,7 +4,7 @@
 
 
   // initiate default values
-     
+
   $change_time = time();  // set time ticket opened
   if( $dtime ) {
     $dtime = $zen->dateParse($dtime);
@@ -12,7 +12,7 @@
   if( $stime ) {
      $stime = $zen->dateParse($stime);
   }
-  
+
   // $description = nl2br(htmlspecialchars($description));
 
   $fields = array(
@@ -50,12 +50,12 @@
     if( !$res ) {
       $errs[] = tr("System Error").": ".tr("Contact could not be edited.")." ".$zen->db_error;
     } else {
-      
-      $agree_id = $id ;    
+
+      $agree_id = $id ;
       $fields = array(
 		      "agree_id"    => "int",
 		      );
-      
+
       $params = array();
       // create an array of existing fields
       foreach(array_keys($fields) as $f) {
@@ -67,15 +67,15 @@
       if( !$res ) {
 	$errs[] = tr("System Error").": ".tr("Contact could not be edited.")." ".$zen->db_error;
       }
-      
+
     }
   }
 
   if( !$errs ) {
-  
-    include("../agreement.php");//test set contact(s)
+
+    include("$rootWWW/agreement.php");//test set contact(s)
     exit;
-    
+
   } else {
     $page_title = tr("Contacts");
     $page_section = "Contacts";
diff -Naur zentrack_2-5-5-1/includes/editCustomSubmit.php zentrack_2-5-5-2/includes/editCustomSubmit.php
--- zentrack_2-5-5-1/includes/editCustomSubmit.php	2005-05-01 23:59:35.000000000 -0700
+++ zentrack_2-5-5-2/includes/editCustomSubmit.php	2005-05-15 16:44:30.000000000 -0700
@@ -6,9 +6,9 @@
   $required = array();
   // the TODO is saved, even if the save failed
   $TODO = 'SAVED';
-                                
+  
   include("$libDir/validateFields.php");
-
+  
   if( !$errs ) {
     if( count($varfield_params) ) {
       // update the ticket info
@@ -20,6 +20,8 @@
       $errs[] = tr("System Error").": ".tr("Ticket could not be edited.")." ".$zen->db_error;
     }
   }
+  
+  $setmode = 'custom';
 
   if( !$errs ) {
     add_system_messages(tr("Variable fields updated: ")." $id.");
@@ -34,6 +36,7 @@
     exit;
   } else {
      include_once("$libDir/nav.php");
+     $zen->printErrors($errs);
      if( $zen->inProjectTypeIDs($bin_id) )
        include("$templateDir/projectView.php");
      else
diff -Naur zentrack_2-5-5-1/includes/footer.php zentrack_2-5-5-2/includes/footer.php
--- zentrack_2-5-5-1/includes/footer.php	2005-05-01 23:59:35.000000000 -0700
+++ zentrack_2-5-5-2/includes/footer.php	2005-05-15 16:44:30.000000000 -0700
@@ -55,6 +55,7 @@
      $debug_text .= "<span class='note'>\n";
      $debug_text .= "HTTP_USER_AGENT: $HTTP_USER_AGENT<br>\n";
      $debug_text .= "SCRIPT_NAME: $SCRIPT_NAME<br>\n";
+     $debug_text .= "QUERY_STRING: $QUERY_STRING<br>\n";
      $debug_text .= "HTTP_HOST: (".(preg_match("/$HTTP_HOST/",$rootUrl)? 'matches rootUrl' : 
         '<b><span class="error">DOES NOT MATCH $rootUrl!!!</span></b>').")<br>\n";
      $debug_text .= "HTTP_COOKIE: $HTTP_COOKIE<br>\n";
diff -Naur zentrack_2-5-5-1/includes/login.php zentrack_2-5-5-2/includes/login.php
--- zentrack_2-5-5-1/includes/login.php	2005-05-01 23:59:35.000000000 -0700
+++ zentrack_2-5-5-2/includes/login.php	2005-05-15 16:44:30.000000000 -0700
@@ -1,90 +1,150 @@
 <?
 if( !ZT_DEFINED ) { die("Illegal Access"); }
 
-  
-  // log user out of zentrack if $logoff
-  if( isset($logoff) && $logoff > 0 ) {
-    $login_id = "";
-    $login_name = "";
-    $login_level = "";
-    $_SESSION['data_groups'] = null;
+$autologin = $zen->settingOn('allow_pwd_save');
+
+// log user out of zentrack if $logoff
+if( isset($logoff) && $logoff > 0 ) {
+  $login_id = "";
+  $login_name = "";
+  $login_level = "";
+  $_SESSION['data_groups'] = null;
+  $autologin = false;
+  setcookie("zentrackKey", "", time());
+}
+
+// auto-login via cookie
+if( $autologin && !$login_id && $zentrackUsername && $zentrackKey ) {
+
+  // make sure the user has this feature turned on
+  $doauto = false;
+  $user = $zen->get_user_by_login($zentrackUsername);
+  if( $user ) {
+    // try to retrieve the user_id so that we can check this users preferences
+    // and determine if their auto-login is enabled.
+    $doauto = $zen->get_prefs( $user['user_id'], 'autologin' );
+    $zen->addDebug("login.php:auto-login", "Checking autlogin setting for user {$user['user_id']}: $doauto", 3);
+  }
+  else {
+    // delete cookie if user is invalid
+    $zen->addDebug("login.php:auto-login", "Invalid username provided(deleting): ".$zen->ffv($zentrackUsername),1);
+    setcookie("zentrackUsername", $zentrackUsername, -1);
   }
 
-  if( !$login_id ) {
-    if( isset($_SERVER['PHP_AUTH_USER']) && isset($_SERVER['PHP_AUTH_PW']) ) {
-      // use htaccess authentication
-      $username = $_SERVER['PHP_AUTH_USER'];
-      $passphrase = $_SERVER['PHP_AUTH_PW'];
+  if( $doauto ) {
+    // automatically log in users based on cookie values
+    $login_id = $zen->login_user($zentrackUsername, $zentrackKey, true);
+    if( $login_id ) {
+      // user login was valid so set session variables
+      $login_level = $zen->user["access_level"];
+      $login_name  = $zen->user["fname"]." ".$zen->user["lname"];
+      $login_inits = $zen->user["initials"];
+      $login_bin   = $zen->user["homebin"];
+      $prefs = $zen->get_prefs($login_id, "language");
+      if(isset($prefs["language"])) { $login_language = $prefs["language"]; }
+      setcookie("zentrackUsername", $zentrackUsername, time()+2592000);
+      setcookie("zentrackKey", $zentrackKey, time()+2592000);
+      $skip = 1;
+      unset($TODO);
+      $zen->addDebug("login.php:auto-login",
+      "User logged in: $login_id,$login_name,$login_level",2);
     }
-    if( isset($username) && $username && isset($passphrase) ) {
-      $login_id = $zen->login_user( $username, $passphrase );
-      if( $login_id && $zen->demo_mode != "on" 
-	  && $zen->user["initials"] != "GUEST" 
+    else {
+      // user login was invalid so get rid of the offending cookie
+      setcookie("zentrackKey", $zentrackKey, -1);
+      $zen->addDebug("login.php:auto-login",
+      "Auto-login was invalid: ".$zen->ffv($zentrackUsername)."||".$zen->ffv($zentrackKey), 1);
+    }
+  }
+}
+
+if( !$login_id ) {
+  if( isset($_SERVER['PHP_AUTH_USER']) && isset($_SERVER['PHP_AUTH_PW']) ) {
+    // use htaccess authentication
+    $username = $_SERVER['PHP_AUTH_USER'];
+    $passphrase = $_SERVER['PHP_AUTH_PW'];
+  }
+  if( isset($username) && $username && isset($passphrase) ) {
+    $login_id = $zen->login_user( $username, $passphrase );
+    if( $login_id && $zen->demo_mode != "on"
+	  && $zen->user["initials"] != "GUEST"
 	  && $zen->encval($zen->user["lname"]) == $zen->encval($passphrase) ) {
-	// this will redirect the user
-	// to a login screen where they can
-	// change their passphrase, since it is 
-	// set to the default
-	$login_level = 'first_login';
-	$login_name  = $zen->user["fname"]." ".$zen->user["lname"];
-	setcookie("zentrackUsername", $username, time()+2592000);
-	$login_inits = $zen->user["initials"];
-	$login_bin   = $zen->user["homebin"];
-	$prefs = $zen->get_prefs($login_id, "language");
-	if(isset($prefs["language"])) { $login_language = $prefs["language"]; }
-	$skip = 1;
-      } else if( $login_id ) {
-	// this will log the user in successfully
-	// and generate session variables, as well
-	// as a cookie to save time logging in
-	// in the future
-	$login_level = $zen->user["access_level"];
-	$login_name  = $zen->user["fname"]." ".$zen->user["lname"];
-	$login_inits = $zen->user["initials"];
-	$login_bin   = $zen->user["homebin"];
-	$prefs = $zen->get_prefs($login_id, "language");
-	if(isset($prefs["language"])) { $login_language = $prefs["language"]; }
-	setcookie("zentrackUsername", $username, time()+2592000);
-	$skip = 1;
-	unset($TODO);
-	$zen->addDebug("login.php:userLogin",
-		       "User logged in: $login_id,$login_name,$login_level",2);
-      } else {
-	// generate an error message and let them try again
-	$msg = "<p>&nbsp;</p><ul><b>".tr("That passphrase didn't work.")."</b></ul>";
+      // this will redirect the user
+      // to a login screen where they can
+      // change their passphrase, since it is
+      // set to the default
+      $login_level = 'first_login';
+      $login_name  = $zen->user["fname"]." ".$zen->user["lname"];
+      setcookie("zentrackUsername", $username, time()+2592000);
+      $login_inits = $zen->user["initials"];
+      $login_bin   = $zen->user["homebin"];
+      $lang_pref = $zen->get_prefs($login_id,'language');
+      if(isset($lang_pref)) { $login_language = $lang_pref; }
+      if( $autologin && $save_my_password ) {
+        // if the user checks the log me in automagically box, we will set a userpref
+        // to do this in the future and create a cookie
+        $zen->update_pref($login_id, 'autologin', 1);
+        setcookie("zentrackKey", $zen->encval($passphrase), time()+2592000);
       }
-      if( $login_bin == -1 ) {
-	$login_bin = null;
+      $skip = 1;
+    } else if( $login_id ) {
+      // this will log the user in successfully
+      // and generate session variables, as well
+      // as a cookie to save time logging in
+      // in the future
+      $login_level = $zen->user["access_level"];
+      $login_name  = $zen->user["fname"]." ".$zen->user["lname"];
+      $login_inits = $zen->user["initials"];
+      $login_bin   = $zen->user["homebin"];
+      $prefs = $zen->get_prefs($login_id, "language");
+      if(isset($prefs["language"])) { $login_language = $prefs["language"]; }
+      setcookie("zentrackUsername", $username, time()+2592000);
+      if( $autologin && $save_my_password ) {
+        // if the user checks the log me in automagically box, we will set a userpref
+        // to do this in the future and create a cookie
+        $zen->update_pref($login_id, 'autologin', 1);
+        setcookie("zentrackKey", $zen->encval($passphrase), time()+2592000);
       }
-    } else {
+      $skip = 1;
+      unset($TODO);
       $zen->addDebug("login.php:userLogin",
-		     "User not logged in, username and passphrase not detected, so creating login form",3);
+      "User logged in: $login_id,$login_name,$login_level",2);
+    } else {
+      // generate an error message and let them try again
+      $msg = "<p>&nbsp;</p><ul><b>".tr("That passphrase didn't work.")."</b></ul>";
     }
-    
-    // user isn't logged in, so show the login form
-     if( !isset($skip) || !$skip ) {
-       // no login has been recieved, but it's required
-       // so generate a login prompt and form
-       if( isset($zentrackUsername) )
-	 $zentrackUsername = strip_tags($zentrackUsername);
-       else
-	 $zentrackUsername = "";
-       $page_title = ucwords(tr("Please log on"));
-       include("$libDir/nav.php");
-       include("$templateDir/loginForm");
-       include("$libDir/footer.php");
-       exit;
-     }          
-
-    // if there is no login id then include the form
-    // or process the form results
-    if( "$login_level" == "first_login" && !ereg("pwc\.php",$SCRIPT_NAME) ) {
-      include_once("$libDir/nav.php");
-      include_once("$templateDir/pwcForm.php");
-      include_once("$libDir/footer.php");
-      exit;    
+    if( $login_bin == -1 ) {
+      $login_bin = null;
     }
+  } else {
+    $zen->addDebug("login.php:userLogin",
+    "User not logged in, username and passphrase not detected, so creating login form",3);
+  }
+
+  // user isn't logged in, so show the login form
+  if( !isset($skip) || !$skip ) {
+    // no login has been recieved, but it's required
+    // so generate a login prompt and form
+    if( isset($zentrackUsername) )
+    $zentrackUsername = strip_tags($zentrackUsername);
+    else
+    $zentrackUsername = "";
+    $page_title = ucwords(tr("Please log on"));
+    include("$libDir/nav.php");
+    include("$templateDir/loginForm");
+    include("$libDir/footer.php");
+    exit;
+  }
+
+  // if there is no login id then include the form
+  // or process the form results
+  if( "$login_level" == "first_login" && !ereg("pwc\.php",$SCRIPT_NAME) ) {
+    include_once("$libDir/nav.php");
+    include_once("$templateDir/pwcForm.php");
+    include_once("$libDir/footer.php");
+    exit;
+  }
 
-  } 
+}
 
 ?>
\ No newline at end of file
diff -Naur zentrack_2-5-5-1/includes/templates/actions/varfield_edit.php zentrack_2-5-5-2/includes/templates/actions/varfield_edit.php
--- zentrack_2-5-5-1/includes/templates/actions/varfield_edit.php	2005-05-01 23:59:39.000000000 -0700
+++ zentrack_2-5-5-2/includes/templates/actions/varfield_edit.php	2005-05-15 16:44:31.000000000 -0700
@@ -1,5 +1,4 @@
 <?
-if( !ZT_DEFINED ) { die("Illegal Access"); }
-
-  include("$libDir/templates/ticket_systemBox.php");
+  if( !ZT_DEFINED ) { die("Illegal Access"); }
+  include("$libDir/templates/ticket_customBox.php");
 ?>
\ No newline at end of file
diff -Naur zentrack_2-5-5-1/includes/templates/agreement_box.php zentrack_2-5-5-2/includes/templates/agreement_box.php
--- zentrack_2-5-5-1/includes/templates/agreement_box.php	2005-05-01 23:59:37.000000000 -0700
+++ zentrack_2-5-5-2/includes/templates/agreement_box.php	2005-05-15 16:44:31.000000000 -0700
@@ -2,57 +2,58 @@
 if( !ZT_DEFINED ) { die("Illegal Access"); }
 
 
-/* 
+/*
 *Show the contacts that are connected to a company
 */
 //echo($company_id);
 if ($company_id>"0") {
-$company = $zen->get_contact($company_id,"ZENTRACK_COMPANY","company_id");	
+$company = $zen->get_contact($company_id,"ZENTRACK_COMPANY","company_id");
 }
 
 if (is_array($company)) {
 
-$name ="<A HREF='".$rootUrl."/contact.php?cid=".$company['company_id']."'>".ucfirst($company["title"])." ".ucfirst($company["office"])."</A>";	
+$name ="<A HREF='".$rootUrl."/contact.php?cid=".$company['company_id']."'>".ucfirst($company["title"])." ".ucfirst($company["office"])."</A>";
 }
 
 ?>
    <table cellpadding="0" cellspacing="0" border="0">
   <tr>
    <td class="ticketCell">
-   <table align="center" width='570' border="0">     
+   <table align="center" width='570' border="0">
      <tr>
        <td valign="top"><table border="0"
           width="430" cellpadding="0" cellspacing="1">
-          
+
     <tr>
-	   <td class="titleCell" colspan="4"><p align="center"><?=$contractnr?></p></td>  
+	   <td class="titleCell" colspan="4"><p align="center"><?=$contractnr?></p></td>
 	  </tr>
 	  <tr>
 	   <td class="smallTitleCell" colspan="2" ><?=uptr("Info")?></td>
 	   <td class="smallTitleCell"  colspan="2" width="50%"><?=uptr("Dates")?></td>
 	  </tr>
 	  <tr>
-	   <td class="small" width="20%" colspan="2"><p align="center"><?=$title?></p></td> 
-	   <td class="small" width="20%"><?=tr("Start Date")?>:</td> 	   
+     <td class="small" width="20%"><?=tr("Title")?>:</td>
+	   <td class="small" width="30%"><?=$title?></td>
+	   <td class="small" width="20%"><?=tr("Start Date")?>:</td>
 	   <td class="small" width="30%"><?if($stime){echo $zen->showDate($stime);}?></td>
 	  </tr>
 	  <tr>
-	   <td class="small" width="20%"><?=tr("Company")?>:</td> 
-	   <td class="small" width="30%"><?=$name?></td>  
-	   <td class="small" width="20%"><?=tr("Expiration Date")?>:</td>  
+	   <td class="small" width="20%"><?=tr("Company")?>:</td>
+	   <td class="small" width="30%"><?=$name?></td>
+	   <td class="small" width="20%"><?=tr("Expiration Date")?>:</td>
 	   <td class="small" width="30%"><?if($dtime){echo $zen->showDate($dtime);}?></td>
 	  </tr>
 	  <tr>
-	   <td class="small" width="20%"></td> 
-	   <td class="small" width="30%"></td> 
-	   <td class="small" width="20%"></td> 
-	   <td class="small" width="30%"></td> 
+	   <td class="small" width="20%"></td>
+	   <td class="small" width="30%"></td>
+	   <td class="small" width="20%"></td>
+	   <td class="small" width="30%"></td>
 	  </tr>
 <?
  if(!empty($description)) {
-?>	  
+?>
 	  <tr>
-	   <td class="smallTitleCell" colspan="4"><?=uptr("Description")?></td>  
+	   <td class="smallTitleCell" colspan="4"><?=uptr("Description")?></td>
 	  </tr>
 	  <tr>
 	   <td class="small" colspan="4"><?=(get_magic_quotes_runtime())?nl2br(stripslashes($description)):nl2br($description); ?></td>
@@ -67,13 +68,13 @@
 $items = $zen->get_contacts($parms,"ZENTRACK_AGREEMENT_ITEM",$sort);
 ?>
 <tr>
-	   <td class="smallTitleCell" colspan="4"><?=uptr("Items")?></td> 
+	   <td class="smallTitleCell" colspan="4"><?=uptr("Items")?></td>
 </tr>
-<tr><td colspan="4"> 
+<tr><td colspan="4">
 <table>
 <?
 if (is_array($items)) {
-  
+
   $class = '';
   foreach($items as $t) {
     $class = $class == 'bars'? 'cell' : 'bars';
@@ -86,23 +87,23 @@
     <td height="25" width="50%" align="middle" >
     <?=strtolower($t["description1"])?>
     </td>
-    </tr>   
-    <?   
-  } 
-  
+    </tr>
+    <?
+  }
+
 } else {
   echo "<tr><td colspan='4'>No items are set</td></tr>" ;
 }?>
 </table>
-</td</tr>  	
+</td</tr>
 
 <?//end items
 ?>
 	 </table>
-	 
+
 	 </td>
    <td valign="top" width='75'>
-       
+
 <table width="120" cellpadding="0" cellspacing="0" border="0">
 <?
 print "<tr>\n<form name='edit_form' action='$rootUrl/actions/agreement_edit.php'>\n";
@@ -134,7 +135,7 @@
 print "<tr>\n<form name='delete_form' action='$rootUrl/actions/agreement_delete.php'>\n";
 print "<td>\n";
 print "<input type='submit' class='actionButtonContact'  value='".uptr('delete')."'";
-print " onClick='return confirm(\"Are you sure you want to permanently delete this contact\")'";
+print " onClick='return confirm(\"".tr("Are you sure you want to permanently delete this agreement?")."\")'";
 print ">\n";
 print "<input type='hidden' name='id' value='$agree_id'>\n";
 print "</td>\n</form>\n</tr>\n";
diff -Naur zentrack_2-5-5-1/includes/templates/fieldMapForm.php zentrack_2-5-5-2/includes/templates/fieldMapForm.php
--- zentrack_2-5-5-1/includes/templates/fieldMapForm.php	2005-05-01 23:59:37.000000000 -0700
+++ zentrack_2-5-5-2/includes/templates/fieldMapForm.php	2005-05-15 16:44:31.000000000 -0700
@@ -13,7 +13,7 @@
    $str = "<a href='$rootUrl/help/find.php?s=admin&p=fieldmap'>".tr('Documentation')."</a>";
    print tr("Please refer to the ? before using this feature", array($str));
  ?></p>
- 
+
 <form method='post' action='<?=$SCRIPT_NAME?>' name='fieldMapForm'>
 <input type='hidden' name='TODO' value='save'>
 <input type='hidden' name='view' value='<?=$zen->ffv($view)?>'>
@@ -88,7 +88,7 @@
   // and get special properties for fields
   $fprops = getFmFieldProps($view, ZenFieldMap::fieldName($f));
   $tprops = $typeprops[$field['field_type']];
-  
+
   // generate row of information
   $s = $field['is_visible']? 'bold' : 'disabled';
   $class = $field['field_type'] == 'section'? "altBars $s" : "bars $s";
@@ -164,9 +164,11 @@
       //if( $f == 'custom_menu1' ) { Zen::printArray($choices); }
       if( is_array($choices) && count($choices) ) {
         $txt = "<select ".fmfName($f,'default_val').">";
-        $txt .= "<option value=''>--</option>";
+        if( $view != 'search_form' ) {
+          $txt .= "<option value=''>--</option>";
+        }
         foreach($choices as $k=>$v) {
-          $sel = $field['default_val'] == $k? " selected" : "";
+          $sel = $field['default_val'] == $k && strlen($field['default_val']) == strlen($k)? " selected" : "";
           $txt .= "<option value='$k'{$sel}>$v</option>";
         }
         $txt .= "</select>";
@@ -177,7 +179,7 @@
       }
     }
     fmfRow($txt,$class);
-    
+
     // field type, not useful for fields which only have label as type
     // or for sections
     if( count($fprops['types']) == 1 && $fprops['types'][0] == 'label' ) {
@@ -211,7 +213,7 @@
     }
     else { fmfRow('1',$class); }
   }
-  
+
   print "</tr>\n";
   $fcount++;
   $prevSection = $field['field_type'] == 'section';
@@ -294,7 +296,7 @@
 function moveRowUp( tdCell ) {
   var v1, v2;
   var thisRow = tdCell.parentNode;
-  quickHighlightRow( thisRow, 'subTitle' );  
+  quickHighlightRow( thisRow, 'subTitle' );
   var previousRow = thisRow.previousSibling;
   var parentNode = thisRow.parentNode;
   while( previousRow && previousRow.nodeName != "TR" ) {
@@ -317,7 +319,7 @@
   v1=parseFloat(document.fieldMapForm[ "orderset[" + thisRow.id + "]" ].value);
   v2=parseFloat(document.fieldMapForm[ "orderset[" + previousRow.id + "]" ].value);
   document.fieldMapForm[ "orderset[" + thisRow.id + "]" ].value=v2;
-  document.fieldMapForm[ "orderset[" + previousRow.id + "]" ].value=v1;  
+  document.fieldMapForm[ "orderset[" + previousRow.id + "]" ].value=v1;
 }
 
 function quickHighlightRow( parentObj, s ) {
@@ -334,14 +336,14 @@
       obj.setAttribute('class', obj.getAttribute('class') + ' ' + s);
       setTimeout("var x = document.getElementById('"+objName+"'); x.setAttribute('class', x.getAttribute('class').substr(0,x.getAttribute('class').indexOf(' "+s+"')));",500);
     }
-    
+
   }
 }
 
 function moveRowDown( tdCell ) {
   var v1, v2;
   var thisRow = tdCell.parentNode;
-  quickHighlightRow( thisRow, 'subTitle' );  
+  quickHighlightRow( thisRow, 'subTitle' );
   var nextRow = thisRow.nextSibling;
   var parentNode = thisRow.parentNode;
   //while( nextRow && nextRow.nodeName != "TR" ) {
@@ -363,29 +365,29 @@
     v2=parseFloat(document.fieldMapForm[ "orderset[" + nextRow.id + "]" ].value);
     document.fieldMapForm[ "orderset[" + thisRow.id + "]" ].value=v2;
     document.fieldMapForm[ "orderset[" + nextRow.id + "]" ].value=v1;
-  } 
-  parentNode.insertBefore(nextRow, thisRow);  
+  }
+  parentNode.insertBefore(nextRow, thisRow);
 }
 
 function toggleRowColor( checkBox ) {
     // navigate to the TR tag for this row
     var trTag = checkBox.parentNode.parentNode;
-    
+
     for(var i=0; i < trTag.childNodes.length; i++) {
       var obj = trTag.childNodes[i];
       if( obj.nodeName != "TD" ) { continue; }
       // perform the class switch
       if( obj.className ) {
         // determine what we are setting the style to based on checkbox
-        obj.className = checkBox.checked? 
+        obj.className = checkBox.checked?
           obj.className.substr(0, obj.className.indexOf(" ")) + " bold" :
           obj.className.substr(0, obj.className.indexOf(" ")) + " disabled";
       }
       else {
         // determine what we are setting the style to based on checkbox
         var oldStyle = obj.getAttribute('class');
-        obj.setAttribute('class', checkBox.checked? 
-          oldStyle.substr(0, oldStyle.indexOf(" "))+" bold" : 
+        obj.setAttribute('class', checkBox.checked?
+          oldStyle.substr(0, oldStyle.indexOf(" "))+" bold" :
           oldStyle.substr(0, oldStyle.indexOf(" "))+" disabled" );
       }
     }
@@ -405,7 +407,7 @@
     }
     return true;
   }
-  
+
   function getNamePrefix( name ) {
     return name.substr(0, name.indexOf('['))
   }
diff -Naur zentrack_2-5-5-1/includes/templates/listTickets.php zentrack_2-5-5-2/includes/templates/listTickets.php
--- zentrack_2-5-5-1/includes/templates/listTickets.php	2005-05-01 23:59:37.000000000 -0700
+++ zentrack_2-5-5-2/includes/templates/listTickets.php	2005-05-15 16:44:31.000000000 -0700
@@ -161,7 +161,14 @@
           <input type="hidden" name="search_text" value="<?=$zen->ffv($search_text)?>">
           <?
            foreach($search_params as $k=>$v) {
-             print "<input type='hidden' name='search_params[$k]' value='".$zen->ffv($v)."'>\n";
+             if( is_array($v) ) {
+               foreach($v as $val) {
+                 print "<input type='hidden' name='search_params[$k][]' value='".$zen->ffv($val)."'>\n";
+               }
+             }
+             else {
+               print "<input type='hidden' name='search_params[$k]' value='".$zen->ffv($v)."'>\n";
+             }
            }
            foreach($search_dates as $k=>$v) {
              print "<input type='hidden' name='search_dates[$k]' value='".$zen->ffv($v)."'>\n";
@@ -169,6 +176,9 @@
            foreach($search_fields as $k=>$v) {
              print "<input type='hidden' name='search_fields[$k]' value='".$zen->ffv($v)."'>\n";
            }
+           if( $or_higher ) {
+             print "<input type='hidden' name='or_higher' value='1'>\n";
+           }
            ?>
      </form>
      <form method="post" action="exportSearch.php" style="display: inline; margin: 0px;">
@@ -176,7 +186,14 @@
           <input type="hidden" name="search_text" value="<?=$zen->ffv($search_text)?>">
           <?
            foreach($search_params as $k=>$v) {
-             print "<input type='hidden' name='search_params[$k]' value='".$zen->ffv($v)."'>\n";
+             if( is_array($v) ) {
+               foreach($v as $val) {
+                 print "<input type='hidden' name='search_params[$k][]' value='".$zen->ffv($val)."'>\n";
+               }
+             }
+             else {
+               print "<input type='hidden' name='search_params[$k]' value='".$zen->ffv($v)."'>\n";
+             }
            }
            foreach($search_dates as $k=>$v) {
              print "<input type='hidden' name='search_dates[$k]' value='".$zen->ffv($v)."'>\n";
@@ -184,6 +201,9 @@
            foreach($search_fields as $k=>$v) {
              print "<input type='hidden' name='search_fields[$k]' value='".$zen->ffv($v)."'>\n";
            }
+           if( $or_higher ) {
+             print "<input type='hidden' name='or_higher' value='1'>\n";
+           }
          ?>
         </nobr>
        </form>
diff -Naur zentrack_2-5-5-1/includes/templates/loginForm zentrack_2-5-5-2/includes/templates/loginForm
--- zentrack_2-5-5-1/includes/templates/loginForm	2005-03-11 07:37:38.000000000 -0800
+++ zentrack_2-5-5-2/includes/templates/loginForm	2005-05-15 16:44:31.000000000 -0700
@@ -8,7 +8,7 @@
 
   <p><b><i style="color:red"><?tr("Login and passphrase are case sensitive!")?></i></b></p>
 
-  <form method="post" action="<?=$SCRIPT_NAME?><?=($id)? "?id=".strip_tags($id) : ""?>">
+  <form method="post" action="<?=$SCRIPT_NAME?><?=($id)? "?id=".$zen->checkNum($id) : ""?>">
 
   <input type=text name=username value='<?=strip_tags($zentrackUsername)?>' width=20 maxlength=20>&nbsp;<?=tr("Login Name")?>
 
@@ -17,8 +17,14 @@
   <input type="password" name="passphrase" width=20 maxlength=20>&nbsp;<?=tr("Password")?>
 
   <p>
+  
+  <? if( $zen && $zen->settingOn('allow_pwd_save') ) { ?>
+  <p>
+  <input type='checkbox' name='save_my_password' value='1'> <?=tr("In the future, log me in automatically (using a cookie)")?>
+  </p>
+  <?  }  ?>
 
-  <input type=hidden name=redirect value='<?=strip_tags($redirect)?>'>
+  <input type=hidden name=redirect value='<?=$zen->ffv($redirect)?>'>
 
   <input type=submit name=TODO value="<?=tr("Log On")?>">
 
diff -Naur zentrack_2-5-5-1/includes/templates/newAgreementForm.php zentrack_2-5-5-2/includes/templates/newAgreementForm.php
--- zentrack_2-5-5-1/includes/templates/newAgreementForm.php	2005-05-01 23:59:38.000000000 -0700
+++ zentrack_2-5-5-2/includes/templates/newAgreementForm.php	2005-05-15 16:44:31.000000000 -0700
@@ -9,9 +9,9 @@
           elem.checked = elem.checked? false : true;
         }
       }
-      if( elem.parentNode ) { 
-        elem.parentNode.parentNode.oldStyle = elem.checked? 'invalidCell' : 'cell'; 
-      }    
+      if( elem.parentNode ) {
+        elem.parentNode.parentNode.oldStyle = elem.checked? 'invalidCell' : 'cell';
+      }
     }
   }
 </script>
@@ -21,28 +21,28 @@
 <input type='hidden' name='TODO' value='submit_form'>
 <?
 if(isset($creator_id)) { ?>
-<input type="hidden" name="creator_id" value="<?=$zen->ffv($creator_id)?>">	
+<input type="hidden" name="creator_id" value="<?=$zen->ffv($creator_id)?>">
 <?
 }
 if(isset($create_time)) { ?>
-<input type="hidden" name="create_time" value="<?=$zen->ffv($create_time)?>">	
+<input type="hidden" name="create_time" value="<?=$zen->ffv($create_time)?>">
 <?
 }
 ?>
-  
+
 <table width="640" align="left" cellpadding="2" cellspacing="2" bgcolor="<?=$zen->settings["color_background"]?>" border="0">
 <tr>
   <td colspan="2" width="640" class="titleCell" align="center">
      <?=tr("Agreement Information")?>
   </td>
 </tr>
-  
+
 <tr>
   <td colspan="2" class="subTitle">
     <?=tr("Global")?>
   </td>
 </tr>
-  
+
 
 <tr>
   <td class="bars" >
@@ -52,7 +52,7 @@
     <input type="text" name="contractnr" size="20" maxlength="25"
       value="<?=$zen->ffv($contractnr)?>">
   </td>
-</tr>  
+</tr>
 <?
 $company = $zen->get_contact_all();
 	if (is_array($company)) {
@@ -75,8 +75,8 @@
 	</td>
 	</tr>
 	<?
-	}   
- ?>   
+	}
+ ?>
 <tr>
   <td class="bars" >
     <?=tr("Title")?>:
@@ -85,7 +85,7 @@
     <input type="text" name="title" size="30" maxlength="50"
 value="<?=$zen->ffv($title)?>">
   </td>
-</tr>          
+</tr>
 
 <tr>
   <td class="bars" >
@@ -94,7 +94,7 @@
   <td class="bars">
     <input type="text" name="stime" size="12" maxlength="10"
 value="<?=($stime)?$zen->showDate($zen->ffv($stime)):""?>">
-    <img name="date_button" src='<?=$rootUrl?>/images/cal.gif' 
+    <img name="date_button" src='<?=$rootUrl?>/images/cal.gif'
   onClick="popUpCalendar(this,document.agreementForm.stime, '<?=$zen->popupDateFormat()?>')"
   alt="Select a Date">
     &nbsp;(<?=tr("optional")?>)
@@ -108,7 +108,7 @@
   <td class="bars">
     <input type="text" name="dtime" size="12" maxlength="10"
 value="<?=($dtime)?$zen->showDate($dtime):""?>">
-    <img name="date_button" src='<?=$rootUrl?>/images/cal.gif' 
+    <img name="date_button" src='<?=$rootUrl?>/images/cal.gif'
       onClick="popUpCalendar(this,document.agreementForm.dtime, '<?=$zen->popupDateFormat()?>')"
       alt="Select a Date">
     &nbsp;(<?=tr("optional")?>)
@@ -119,17 +119,17 @@
     <?=tr("Description")?>:
   </td>
 </tr>
-  
+
 <tr>
   <td colspan="2" class="bars">
-    <textarea cols="60" rows="5" name="description"><?= 
+    <textarea cols="60" rows="5" name="description"><?=
 	    $zen->ffv($description)
     ?></textarea>
   </td>
 </tr>
 
 
-	     
+
 <tr>
   <td class="titleCell" colspan="2" align="center">
   <?=tr("Click button to")?> <?=($td)? tr("save your changes"):tr("create your agreement")?>.
@@ -158,17 +158,17 @@
 	  $parms = array(1 => array(1 => "agree_id", 2 => "=", 3 => $id)
 		);
   }
-  
+
 	$sort = "item_id asc";
 	$contacts = $zen->get_contacts($parms,"ZENTRACK_AGREEMENT_ITEM",$sort);
 
   if (is_array($contacts)) {
-	 //print_r($contacts); 
-	
+	 //print_r($contacts);
+
 	 foreach($contacts as $t) {
       ?>
    <tr class='cell'
-       onmouseover='mClassX(this, "altCellInv", "hand")' 
+       onmouseover='mClassX(this, "altCellInv", "hand")'
        onmouseout='mClassX(this)'
        onclick='checkMyBox("drops_<?=$t['item_id']?>", event)'>
    <td height="25" width="20" valign="middle">
@@ -180,28 +180,28 @@
    <td height="25"  width="400" valign="middle">
    <pre><?=$zen->ffv($t["description1"])?></pre>
    </td>
-   <td><input id='drops_<?=$t['item_id']?>' 
-          type='checkbox' name='drops[]' 
-          value='<?=$t['item_id']?>'></td>
-   </tr>   
-   <?   
+   <td><input id='drops_<?=$t['item_id']?>'
+          type='checkbox' name='drops[]'
+          value='<?=$zen->ffv($t['item_id'])?>'></td>
+   </tr>
+   <?
    } ?>
    <tr>
 	<td colspan="2">
-         <input type="submit" 
-	  value=" <?=tr("Drop Items")?> " 
+         <input type="submit"
+	  value=" <?=tr("Drop Items")?> "
           class="actionButton"
           onClick='return rerouteAgreementForm("removeItems")'
-         > 
+         >
 	</td>
 	</tr>
 	<?
   } else {
 	 echo "No items are set" ;
   }
-  
+
   ?>
- 
+
    </table>
   </td>
 </tr>
@@ -214,16 +214,16 @@
 <tr>
 <td class="bars" colspan="2">
 	<?=tr("Name")?>:
-    <input type="text" name="name1" size="30" maxlength="40" value="<?=strip_tags($name1)?>">
+    <input type="text" name="name1" size="30" maxlength="40" value="<?=$zen->ffv($name1)?>">
 
 	<?=tr("Description")?>:
-    <textarea cols="22" rows="2" name="description1"><?=stripslashes($description1)?></textarea>
+    <textarea cols="22" rows="2" name="description1"><?=$zen->ffv($description1)?></textarea>
 
 	</td>
 </tr>
 <tr>
 <td>
-<input type="submit" value=" <?=tr("Add Item")?> " 
+<input type="submit" value=" <?=tr("Add Item")?> "
    onClick='return rerouteAgreementForm("addItems")'
    class="actionButton">
 </td>
diff -Naur zentrack_2-5-5-1/includes/templates/optionsMenu.php zentrack_2-5-5-2/includes/templates/optionsMenu.php
--- zentrack_2-5-5-1/includes/templates/optionsMenu.php	2005-05-01 23:59:38.000000000 -0700
+++ zentrack_2-5-5-2/includes/templates/optionsMenu.php	2005-05-15 16:44:31.000000000 -0700
@@ -5,11 +5,17 @@
   <b><?=tr("Account Options")?></b>
   <ul>
     <b><a href="<?=$rootUrl?>/misc/pwc.php"><?=tr("Change Password")?></a></b>
-    <? if( $zen->settings["allow_pwd_save"] == "on" ) { ?>
-    <br><b><a href="<?=$rootUrl?>/misc/cookie.php"><?=tr("Remember My Password")?></a></b>
-    <? } ?>
-    <br><b><a href="<?=$rootUrl?>/misc/homebin.php"><?=tr("Change Default Bin")?></a></b>
-    <br><b><a href="<?=$rootUrl?>/misc/language.php"><?=tr("Change Language")?></a></b>
+    <? if( $zen->settings["allow_pwd_save"] == "on" ) {
+         if( $zen->get_prefs($login_id,'autologin') ) {
+           print "<p><b><a href='$rootUrl/misc/autologin.php?setauto=off'>".tr("Disable Auto-Login")."</a></b>";
+         }
+         else {
+           print "<p><b><a href='$rootUrl/misc/autologin.php?setauto=on'>".tr("Enable Auto-Login")."</a></b>";
+         } 
+       }
+     ?>
+    <p><b><a href="<?=$rootUrl?>/misc/homebin.php"><?=tr("Change Default Bin")?></a></b>
+    <p><b><a href="<?=$rootUrl?>/misc/language.php"><?=tr("Change Language")?></a></b>
   </ul>
 </ul>
 
diff -Naur zentrack_2-5-5-1/includes/templates/searchResults.php zentrack_2-5-5-2/includes/templates/searchResults.php
--- zentrack_2-5-5-1/includes/templates/searchResults.php	2005-05-01 23:59:38.000000000 -0700
+++ zentrack_2-5-5-2/includes/templates/searchResults.php	2005-05-15 16:44:31.000000000 -0700
@@ -26,7 +26,7 @@
   // organize the search params
   if( is_array($search_params) ) {
     foreach($search_params as $k=>$v) {
-      if( !strlen($v) || is_array($v) && count($v) == 0 ) { continue; }
+      if( !strlen($v) || (is_array($v) && count($v) == 0) ) { continue; }
       if( is_array($v) && count($v) == 1 && !strlen($v[0]) ) { continue; }
       
       $props = getFmFieldProps($view, $k);
@@ -36,7 +36,7 @@
       $type = $props['data_type'];
       $zen->addDebug('searchResults.php', "Including search_param[$k] ($type)", 3);
       if( $type == 'boolean' && $field['num_rows'] == 1 ) {
-        $params[] = array($k, ($v? "=":"!="), 1, 1);
+        $params[] = array($k, ($v? "=":"<"), 1, 1);
         continue;
       }
       else if( $type == 'date' ) {
@@ -95,14 +95,13 @@
             break;
           case "bin_id":
             if( is_array($v) ) {
-              $ok = true;
+              $binvals = array();
               foreach($v as $val) {
-                if( !in_array($val, $userBins) ) {
-                  $ok = false;
-                  break;
+                if( in_array($val, $userBins) ) {
+                  $binvals[] = $val;
                 }
               }
-              $params[] = array($k, 'IN', $v, 1);
+              $params[] = array($k, 'IN', $binvals, 1);
             }
             else if( in_array($v, $userBins) ) {
               $params[] = array($k, '=', $v, 1);
diff -Naur zentrack_2-5-5-1/includes/templates/ticketView.php zentrack_2-5-5-2/includes/templates/ticketView.php
--- zentrack_2-5-5-1/includes/templates/ticketView.php	2005-05-01 23:59:38.000000000 -0700
+++ zentrack_2-5-5-2/includes/templates/ticketView.php	2005-05-15 16:44:31.000000000 -0700
@@ -14,7 +14,6 @@
   **  extracted before including this page.
   */
 ?>
- 
  <table width="640" cellspacing="1" cellpadding="2">
   <tr>
     <td width="80" valign="top"><? include("$templateDir/ticket_actionBar.php"); ?></td>
@@ -23,4 +22,4 @@
   <tr>
     <td colspan="2" width="640" valign="top"><? include("$templateDir/ticket_box.php"); ?></td>
   </tr>
-</table>
+</table>
\ No newline at end of file
diff -Naur zentrack_2-5-5-1/includes/templates/ticket_actionBar.php zentrack_2-5-5-2/includes/templates/ticket_actionBar.php
--- zentrack_2-5-5-1/includes/templates/ticket_actionBar.php	2005-05-01 23:59:38.000000000 -0700
+++ zentrack_2-5-5-2/includes/templates/ticket_actionBar.php	2005-05-15 16:44:31.000000000 -0700
@@ -46,8 +46,7 @@
       print "<input type='hidden' name='setmode' value='$a'>\n";
       print "</td>\n</form>\n</tr>\n";
     }
-  }
-  
+  }  
 }?>
 </table>
 
diff -Naur zentrack_2-5-5-1/includes/templates/ticket_box.php zentrack_2-5-5-2/includes/templates/ticket_box.php
--- zentrack_2-5-5-1/includes/templates/ticket_box.php	2005-05-01 23:59:38.000000000 -0700
+++ zentrack_2-5-5-2/includes/templates/ticket_box.php	2005-05-15 16:44:31.000000000 -0700
@@ -55,6 +55,7 @@
 
   $vx = $zen->getCustomFields(1, $page_type, 'Custom');
   $counts = $zen->get_ticket_stats($id);
+    
   foreach( $tabs as $key=>$t ) {
     if( $key == 'custom' && (!$vx || !count($vx)) ) {
       // only display the custom tab if there are fields on it
@@ -76,17 +77,15 @@
     
     // Possible translations: (for the benefit of translation maintenance tools)
     // tr("Related"); tr("Notify"); tr("Tasks");
-    $txt = (isset($counts[$key]) && $counts[$key])?
-      $t." (".$counts["$key"].")" : $t;
+    $txt = (isset($counts[$key]) && $counts[$key])? $t." (".$counts["$key"].")" : $t;
     $w = ($key == "attachments")? 85 : 60;
-    print "<td class='$class' height='$height_num' width='$w'>";
-    print "<a href='$pageUrl?id=$id&setmode=$key' class='$lclass'>$txt</a></td>\n";
+    print "<td class='{$class}' height='{$height_num}' width='{$w}'>";
+    print "<a href='{$pageUrl}?id={$id}&setmode={$key}' class='{$lclass}'>{$txt}</a></td>\n";
     if( $i < count($tabs) ) {
-      print "<td width='3'><img src='$rootUrl/images/empty.gif' width='3' height='1'></td>\n";
+      print "<td width='3'><img src='{$rootUrl}/images/empty.gif' width='3' height='1'></td>\n";
     }
     $i++;
   }
-
   print "</tr></table>\n";
 
 ?></td>
diff -Naur zentrack_2-5-5-1/includes/templates/ticket_logBox.php zentrack_2-5-5-2/includes/templates/ticket_logBox.php
--- zentrack_2-5-5-1/includes/templates/ticket_logBox.php	2005-05-08 07:41:03.000000000 -0700
+++ zentrack_2-5-5-2/includes/templates/ticket_logBox.php	2005-05-15 16:44:31.000000000 -0700
@@ -1,64 +1,64 @@
 <? if( !ZT_DEFINED ) { die("Illegal Access"); } ?>
 
-  <table width="600" align="center" cellpadding="2" cellspacing="2">
-      <?
-  $logs = $zen->get_logs($id);
-  if( is_array($logs) && count($logs) > 0) {
-    print "<tr><td class='titleCell'>".tr("Log Entries");
-    print "</td></tr>\n";
-    $att = $zen->get_attachments($id,null,1);
-    $i = 1;
-    $sep = "\n--";
-    foreach($logs as $l) {
-      if( $i == 20 ) {
-	print $title_row;
-	$i = 1;
-      }
-      $lid = $l["lid"];
-      $style = ($style == 'cell')? 'bars' : 'cell';
-      //
-      // the details
-      print "<tr>\n";
-      print "<td class='$style' style='color:".$zen->settings["color_text"]."'>";
-      print $zen->showDateTime($l["created"]);
-      print $sep.$l["action"];
-      print "<br>".$zen->getBinName($l["bin_id"]);
-      print $sep.$zen->formatName($l["user_id"],2);
-      print (strlen($l["hours"]))? $sep.$l["hours"]." hrs":"";
-      //
-      // the log and attachments
-      if( $l["entry"] ) {
-	print "<br>\n";
-	$l["entry"] = nl2br($zen->ffv($l["entry"]));
-	$l["entry"] = preg_replace("#\&amp;#", "&", $l["entry"]);
-//	$l["entry"] = preg_replace("#(https?://[a-zA-Z0-9_/.-]+[a-zA-Z0-9\-_]+\.[a-z]{2,3}(/[a-zA-Z/\._\?=&0-9-]+))#", "<a href='\\1' target='_blank'>\\1</a>", $l["entry"]);
-	$l["entry"] = preg_replace("|(https?://[a-zA-Z0-9_/.-]+(/[a-zA-Z0-9/\.,_\?=&;:#+$!~*%'()-]+[a-zA-Z0-9_=&#+~%-]))|",
-       "<a href='\\1' target='_blank'>\\1</a>", $l["entry"]);
-	$l["entry"] = preg_replace("#([^/])(www\.)([a-zA-Z_/.-]+[a-zA-Z])#", 
+<table width="600" align="center" cellpadding="2" cellspacing="2">
+<?
+$logs = $zen->get_logs($id);
+if( is_array($logs) && count($logs) > 0) {
+  print "<tr><td class='titleCell'>".tr("Log Entries");
+  print "</td></tr>\n";
+  $att = $zen->get_attachments($id,null,1);
+  $i = 1;
+  $sep = "\n--";
+  foreach($logs as $l) {
+    if( $i == 20 ) {
+      print $title_row;
+      $i = 1;
+    }
+    $lid = $l["lid"];
+    $style = ($style == 'cell')? 'bars' : 'cell';
+    //
+    // the details
+    print "<tr>\n";
+    print "<td class='$style' style='color:".$zen->settings["color_text"]."'>";
+    print $zen->showDateTime($l["created"]);
+    print $sep.$l["action"];
+    print "<br>".$zen->getBinName($l["bin_id"]);
+    print $sep.$zen->formatName($l["user_id"],2);
+    print (strlen($l["hours"]))? $sep.$l["hours"]." hrs":"";
+    //
+    // the log and attachments
+    if( $l["entry"] ) {
+      print "<br>\n";
+      $l["entry"] = $zen->ffvText($l["entry"]);
+      //$l["entry"] = preg_replace("#\&amp;#", "&", $l["entry"]);
+      //	$l["entry"] = preg_replace("#(https?://[a-zA-Z0-9_/.-]+[a-zA-Z0-9\-_]+\.[a-z]{2,3}(/[a-zA-Z/\._\?=&0-9-]+))#", "<a href='\\1' target='_blank'>\\1</a>", $l["entry"]);
+      $l["entry"] = preg_replace("|(https?://[a-zA-Z0-9_/.-]+(/[a-zA-Z0-9/\.,_\?=&;:#+$!~*%'()-]+[a-zA-Z0-9_=&#+~%-]))|",
+      "<a href='\\1' target='_blank'>\\1</a>", $l["entry"]);
+      $l["entry"] = preg_replace("#([^/])(www\.)([a-zA-Z_/.-]+[a-zA-Z])#", 
 	    "\\1<a href='http://www.\\3' target='_blank'>www.\\3</a>", $l["entry"]);
-	$l["entry"] = preg_replace("#^(www\.)([a-zA-Z_/.-]+[a-zA-Z])#", 
-	   "<a href='http://www.\\2' target='_blank'>www.\\2</a>", $l["entry"]);
-
- 
-	print $l["entry"];
+      $l["entry"] = preg_replace("#^(www\.)([a-zA-Z_/.-]+[a-zA-Z])#", 
+      "<a href='http://www.\\2' target='_blank'>www.\\2</a>", $l["entry"]);
+      
+      
+      print $l["entry"];
+    }
+    if( is_array($att["$id"]["$lid"]) ) {
+      print "<p><b>".uptr("Attachments").":</b><br>";
+      foreach( $att["$id"]["$lid"] as $a ) {
+        print "<a href='".$zen->settings["url_view_attachment"]
+        ."?aid=$a[attachment_id]' ";
+        print "target='_blank'>$a[name]</a>";
+        print "<span class='small'>($a[description])</span><br>\n";
       }
-      if( is_array($att["$id"]["$lid"]) ) {
-	print "<p><b>".uptr("Attachments").":</b><br>";
-	foreach( $att["$id"]["$lid"] as $a ) {
-	  print "<a href='".$zen->settings["url_view_attachment"]
-	    ."?aid=$a[attachment_id]' ";
-	  print "target='_blank'>$a[name]</a>";
-	  print "<span class='small'>($a[description])</span><br>\n";
-	}
-      }
-      print "</td></tr>\n";
-      print "</td>\n";
-      $i++;
-    }	    
-  } else {
-    print "<tr><td height='300' valign='top'>"
-      .tr("The log is empty")."</td></tr>\n";
-  }
+    }
+    print "</td></tr>\n";
+    print "</td>\n";
+    $i++;
+  }	    
+} else {
+  print "<tr><td height='300' valign='top'>"
+  .tr("The log is empty")."</td></tr>\n";
+}
 ?>
-  </table>
+</table>
 
diff -Naur zentrack_2-5-5-1/includes/templates/ticket_titleBox.php zentrack_2-5-5-2/includes/templates/ticket_titleBox.php
--- zentrack_2-5-5-1/includes/templates/ticket_titleBox.php	2005-05-08 07:41:03.000000000 -0700
+++ zentrack_2-5-5-2/includes/templates/ticket_titleBox.php	2005-05-15 16:44:31.000000000 -0700
@@ -58,5 +58,4 @@
      </tr>
     </table></td>
   </tr>
-</table>
-
+</table>
\ No newline at end of file
diff -Naur zentrack_2-5-5-1/includes/zen.class zentrack_2-5-5-2/includes/zen.class
--- zentrack_2-5-5-1/includes/zen.class	2005-05-08 07:41:02.000000000 -0700
+++ zentrack_2-5-5-2/includes/zen.class	2005-05-15 16:44:30.000000000 -0700
@@ -11,6 +11,55 @@
    // it also extends the zenDate class to pull those
    // functions into use for the rest of zenTrack.
    
+  /**
+   * STATIC: safely check for equality of two values when type is unsure
+   *
+   * This method can accurately compare nulls, values which are not set,
+   * Objects and most other things accurately.
+   *
+   * @param mixed $val1
+   * @param mixed $val2
+   * @return boolean
+   */
+  function safeEquals( $val1, $val2 ) {
+    if( !isset($val1) xor !isset($val2) ) {
+      // if only one is set
+      return false;
+    }
+    else if( is_array($val1) && is_array($val2) ) {      
+      return Zen::arrayEquals($val1, $val2);
+    }
+    else if( $val1 != $val2 || strlen($val1) != strlen($val2) ) {
+      return false;
+    }
+    return true;
+  }
+   
+   
+  /**
+   * STATIC: Recursively checks values of two arrays for equality
+   *
+   * @param array $arr1
+   * @param array $arr2
+   * @return boolean
+   */
+  function arrayEquals( $arr1, $arr2 ) {
+    foreach( $arr1 as $key=>$val ) {
+      if( !isset($arr2[$key]) ) {
+        // don't try to test if arr2[key] isn't set or causes warnings
+        // so see if val isset
+        if( isset($val) ) {
+          return false;
+        }
+        continue;
+      }
+      if( !Zen::safeEquals($val, $arr2[$key]) ) {
+        return false;
+      }
+    }
+    return true;
+  }
+
    /*
    **  TEXT FORMATTING
    */
@@ -81,8 +130,7 @@
    }
    
    function ffvText( $val = '', $strlen = 0 ) {
-     if( $strlen && strlen($val) > $strlen ) { $val = substr($val,0,$strlen); }
-     $val = $this->ffv($val);
+     $val = $this->ffv($val, $strlen);
      $val = nl2br($val);
      return $val;
    }
@@ -90,11 +138,20 @@
    function ffv( $val = '', $strlen = 0 ) {
      // converts special characters to ascii codes
      // to be used in <input> fields
+     if( is_array($val) ) {
+       $vars = array();
+       foreach($val as $k=>$v) {
+         $vars[$k] = $this->ffv($v);
+       }
+       return $vars;
+     }
      if( strlen($val) ) {
        if( $strlen && strlen($val) > $strlen ) { $val = substr($val,0,$strlen); }
-       $val = htmlspecialchars($val,ENT_QUOTES,$this->settings["character_set"]);
-       $val = str_replace('\\', '&#92;', $val);
-       $val = str_replace('&amp;', '&', $val);
+       if( $val ) {
+         $val = htmlspecialchars($val,ENT_QUOTES,$this->settings["character_set"]);
+         $val = str_replace('\\', '&#92;', $val);
+         // $val = str_replace('&amp;', '&', $val);
+       }
      }
      return $val;
    }
@@ -713,7 +770,7 @@
    * @return array an empty array if both values are null or empty
    */
    function mergeArrays($arr1, $arr2) {
-     if( is_array($arr1) && is_array($arr2) ) { return array_merge($arr1,$arr2); }
+     if( is_array($arr1) && is_array($arr2) ) { return $arr1 + $arr2; }
      else if( is_array($arr1) ) { return $arr1; }
      else if( is_array($arr2) ) { return $arr2; }
      else { return array(); }
diff -Naur zentrack_2-5-5-1/includes/zenDate.class zentrack_2-5-5-2/includes/zenDate.class
--- zentrack_2-5-5-1/includes/zenDate.class	2005-05-01 23:59:36.000000000 -0700
+++ zentrack_2-5-5-2/includes/zenDate.class	2005-05-15 16:44:31.000000000 -0700
@@ -303,13 +303,13 @@
      // separator to parse euro date
      $sep = '[/.,-]';
      // if we have eurodate = true and the date is in the format dd/mm/yy[yy] then parse as euro
-     if( $this->euroEnabled && preg_match("#[0-9]{2}{$sep}[0-9]{2}{$sep}[0-9]{2,4}#", $ctime) ) {
+     if( $this->euroEnabled && preg_match("#^[0-9]{2}{$sep}[0-9]{2}{$sep}[0-9]{2,4}#", $ctime) ) {
        // split date and time
        list($date, $time) = explode(' ', $ctime);
        // break up date
        list($day, $month, $year) = split($sep, $date);
        // break up time
-       list($hours, $mins, $seconds) = explode(':', $time);
+       list($hours, $mins, $seconds) = explode('[.,:-]', $time);
        // create a unix timestamp and return
        return mktime($hours, $mins, $seconds, $month, $day, $year);
      }
diff -Naur zentrack_2-5-5-1/includes/zenTemplate.class zentrack_2-5-5-2/includes/zenTemplate.class
--- zentrack_2-5-5-1/includes/zenTemplate.class	2005-05-01 23:59:36.000000000 -0700
+++ zentrack_2-5-5-2/includes/zenTemplate.class	2005-05-15 16:44:31.000000000 -0700
@@ -144,7 +144,7 @@
 	    foreach($vars as $k=>$v) {
         $fv = $this->_getVar("field_value");
 	      $tmp=$str;
-	      if($fv==$k || is_array($fv) && in_array($k, $fv) || 
+	      if($fv==$k && strlen($fv) == strlen($k) || is_array($fv) && in_array($k, $fv) || 
             (!is_array($fv) && !strlen($k) && !strlen($fv))) {
 	        $tmp = str_replace("{selected}", " selected", $tmp);
 	        $tmp = str_replace("{checked}", " checked", $tmp);
diff -Naur zentrack_2-5-5-1/includes/zenTrack.class zentrack_2-5-5-2/includes/zenTrack.class
--- zentrack_2-5-5-1/includes/zenTrack.class	2005-05-01 23:59:36.000000000 -0700
+++ zentrack_2-5-5-2/includes/zenTrack.class	2005-05-15 16:44:31.000000000 -0700
@@ -3162,33 +3162,63 @@
     return( $this->db_result($query) );
   }
 
+  /**
+   * Returns preferences for a user
+   *
+   * @param int $user_id the user to retrieve
+   * @param string $pref if provided, only the value of this preference is returned, otherwise an array is returned
+   * @return mixed a (string)value if $pref is specified, otherwise an array of all prefs for user
+   */
   function get_prefs( $user_id, $pref = '' ) {
     // try to retrieve from session
     $prefSet = $this->_session->getDataType('prefs');
-    if( isset($prefSet) && isset($prefSet["U$user_id"]) ) { return $prefSet["U$user_id"]; }
+    if( isset($prefSet) && isset($prefSet["U$user_id"]) ) { 
+      if( $pref ) {
+        return array_key_exists($pref, $prefSet["U$user_id"])?
+          $prefSet["U$user_id"][$pref] : null;
+      }
+      else { return $prefSet["U$user_id"]; } 
+    }
     if( !isset($prefSet) ) { $prefSet = array(); }
     
     // retrieve from database
     $user_id = $this->checkNum($user_id);
-    $query = "SELECT * FROM ".$this->table_preferences
+    $query = "SELECT prefname,prefval FROM ".$this->table_preferences
       ." WHERE user_id = $user_id";
-    if( $pref ) {
-      $query .= " AND prefname = '$prefname'";
-    }
-    $vals = array();
-    $vars = $this->db_queryIndexed($query);
-    if( is_array($vars) ) {
-      foreach($vars as $v) {
-        $n = $v["prefname"];
-        $vals["$n"] = $v["prefval"];
-      }
-    }
+    $vals = $this->db_listIndexed($query);
     
     // store in session
     $prefSet["U$user_id"] = $vals;
     $this->_session->storeDataType("prefs",$prefSet);
     
-    return $vals;
+    if( $pref ) {
+      return array_key_exists($pref, $vals)?
+        $vals[$pref] : null;
+    }
+    else { return $vals; } 
+  }
+  
+  /**
+   * Update (or add) a single preference in the database
+   *
+   * @param int $user_id
+   * @param string $pref (can only contain [a-zA-Z_ -])
+   * @param string $value the value to be inserted, null or "" should work fine
+   */
+  function update_pref( $user_id, $pref, $value ) {
+    $this->delete_prefs($user_id,$prefname);
+    $this->_session->clearDataType('prefs');
+    $vals = array(
+      'user_id'  => $this->checkNum($user_id),
+      'prefname' => $this->checkAlphaNum($pref, '_ -'),
+      'prefval'  => $value
+    );
+    list($cols,$vals) = $this->makeInsertVals($vals);
+    $query = "INSERT INTO ".$this->table_preferences
+      ." ($cols) VALUES($vals)";
+    $res = $this->db_result($query);
+    $this->addDebug('update_pref', "[$res]".$query, 3);
+    return $res;
   }
 
   function update_prefs( $user_id, $params, $prefname = '' ) {
@@ -3205,7 +3235,7 @@
     foreach($params as $p) {
       $vars = array("user_id"=>$user_id);
       foreach($p as $k=>$v) {
-        $vars["prefname"] = $k;
+        $vars["prefname"] = $this->checkAlphaNum($k,'_ -');
         $vars["prefval"] = strlen($v)? $v : "NULL";
       }
       list($cols,$vals) = $this->makeInsertVals($vars);
@@ -3225,6 +3255,7 @@
     // deletes prefs entries by user_id
     // user_id can be an array      
     $user_id = $this->checkNum($user_id);
+    $prefname = $this->checkAlphaNum($prefname, '_ -');
     $query = "DELETE FROM ".$this->table_preferences." WHERE user_id = $user_id";
     if( $prefname ) {
       $query .= " AND prefname = '$prefname'";
@@ -3366,12 +3397,13 @@
    */
 
 
-  function login_user( $username, $passphrase ) {
+  function login_user( $username, $passphrase, $from_cookie = false ) {
     // perform a login check for username and passphrase
     // returns the user's user_id
     $username = $this->checkAlphaNum($username);
+    $pass = $from_cookie? $this->checkSlashes($passphrase) : $this->checkSlashes($this->encval($passphrase));
     $query = "select user_id from ".$this->table_users
-      ." where login = '$username' and passphrase = '".$this->encval($passphrase)."' and active > 0";
+      ." where login = '$username' and passphrase = $pass and active > 0";
     $user_id = $this->db_get($query);
     $this->addDebug("zentrack.class:login_user($user_id)",$query,2);
     if( $user_id )
@@ -4606,14 +4638,6 @@
                                "egate"    => 1,
                                "level"    => $this->settings["level_view"]
                                );
-      $actions["edit"] = array(
-                               "owner"    => 0,
-                               "access"   => 1,
-                               "status"   => array('OPEN','PENDING'),
-                               "override" => 0,
-                               "egate"    => 0,
-                               "level"    => $this->settings["level_edit"]
-                               );
       $actions["varfield_edit"] = array(
                                         "owner"    => 1,
                                         "access"   => 1, 
@@ -5044,6 +5068,10 @@
     
     return($vals);
   }
+  
+  function settingOn( $setting_name ) {
+    return array_key_exists($setting_name, $this->settings) && $this->settings["$setting_name"] == "on";
+  }
 
   function getTypeName($id) {
     // retrieves the name of the type
@@ -5226,7 +5254,7 @@
     // prevent multiple queries)
     // if $language is given, it will override the value
     // from the configVars file
-
+    
     include("$file");
 
     $this->zen();
diff -Naur zentrack_2-5-5-1/www/actions/action_header.php zentrack_2-5-5-2/www/actions/action_header.php
--- zentrack_2-5-5-1/www/actions/action_header.php	2005-03-11 07:37:39.000000000 -0800
+++ zentrack_2-5-5-2/www/actions/action_header.php	2005-05-15 16:45:02.000000000 -0700
@@ -44,6 +44,9 @@
   if( in_array($basename,array_keys($zen->getActions())) ) {
     $action = $basename;
   }
+  else if( $basename == 'editcustomsubmit' ) {
+    $action = 'varfield_edit';
+  }
   else if( $basename == "editticketsubmit" || $basename == "editsubmit" ) {
     $action = "edit";
   }
diff -Naur zentrack_2-5-5-1/www/actions/agreement_edit.php zentrack_2-5-5-2/www/actions/agreement_edit.php
--- zentrack_2-5-5-1/www/actions/agreement_edit.php	2004-11-22 13:19:04.000000000 -0800
+++ zentrack_2-5-5-2/www/actions/agreement_edit.php	2005-05-15 16:45:03.000000000 -0700
@@ -1,60 +1,59 @@
 <?
   /*
-  **  Action: edit ticket
+  **  Action: edit agreement
   */
   include_once("../header.php");
   $page_title = tr("Agreement #?", array($id));
   $expand_agreement = 1;
-  
-  
+
+
   include("$libDir/nav.php");
-  
+
   //add item button action
-  if ($test=="test") {
-	
-	$description1 = nl2br($zen->ffv($description1));
-	$agree_id = "0";
-	$create_time = time();
-
-  $fields = array(
-		  "agree_id"    => "int",
-		  "description1" => "ignore",
-		  "name1"       => "text",
-		  "create_time" => "int",
-		  );
-		  
-	foreach(array_keys($fields) as $f) {
-	if( strlen($$f) ) {
-	   $params["$f"] = $$f;
-	}
-	}
-	
-	$params["creator_id"] = $login_id;
-   	if (!$id){
-  		$idi = $zen->add_contact($params,"ZENTRACK_AGREEMENT_ITEM");
-		} else {
-			$params["agree_id"] = $id;
-			$idi = $zen->add_contact($params,"ZENTRACK_AGREEMENT_ITEM");
-		}
-  if( !$idi ) {
-	$errs[] = tr("Could not create item.") . " " .$zen->db_error;
-  }
-  $create_time = null;   
-  $idi = null;   
-  $description1 = null;
-  $name1 = null;
+  if ($TODO == "addItems") {
+
+    $agree_id = "0";
+    $create_time = time();
+
+    $fields = array(
+        "agree_id"    => "int",
+        "description1" => "text",
+        "name1"       => "text",
+        "create_time" => "int",
+        );
+
+    foreach(array_keys($fields) as $f) {
+      if( strlen($$f) ) {
+         $params["$f"] = $$f;
+      }
+    }
+
+    $params["creator_id"] = $login_id;
+    if (!$id){
+      $idi = $zen->add_contact($params,"ZENTRACK_AGREEMENT_ITEM");
+    } else {
+      $params["agree_id"] = $id;
+      $idi = $zen->add_contact($params,"ZENTRACK_AGREEMENT_ITEM");
+    }
+    if( !$idi ) {
+      $errs[] = tr("Could not create item.") . " " .$zen->db_error;
+    }
+    $create_time = null;
+    $idi = null;
+    $description1 = null;
+    $name1 = null;
 	}
 	//drop button action
-	if ($test =="test1"){
-		  if( is_array($drops) ) {
-    		// drop items in list
-    		for($i=0; $i<count($drops); $i++) {
-      	// clean up numbers just in case
-      	$n = $zen->checkNum($drops[$i]);
-      	
-				$res = $zen->delete_contact( $n,"ZENTRACK_AGREEMENT_ITEM","item_id");
-				}
-  		}
+	else if ($TODO == "removeItems" ) {
+    if( is_array($drops) ) {
+      // drop items in list
+      for($i=0; $i<count($drops); $i++) {
+      // clean up numbers just in case
+      $n = $zen->checkNum($drops[$i]);
+
+      $res = $zen->delete_contact( $n,"ZENTRACK_AGREEMENT_ITEM","item_id");
+      }
+    }
 	}
 
   include("$templateDir/editAgreementForm.php");
diff -Naur zentrack_2-5-5-1/www/actions/log.php zentrack_2-5-5-2/www/actions/log.php
--- zentrack_2-5-5-1/www/actions/log.php	2003-03-27 13:10:36.000000000 -0800
+++ zentrack_2-5-5-2/www/actions/log.php	2005-05-15 16:45:03.000000000 -0700
@@ -12,21 +12,21 @@
   
   if( $actionComplete == 1 ) {
     $input = array(
-		   "id"       => "int",
-		   "comments" => "html",
-		   "hours"    => "num",
-		   "log_action"   => "text"
-		   );
+    "id"       => "int",
+    "comments" => "text",
+    "hours"    => "num",
+    "log_action"   => "text"
+    );
     $zen->cleanInput($input);
     $required = array("id","action");
     foreach($required as $r) {
       if( !$$r ) {
-	$errs[] = tr(" ? is required", array($r));
+        $errs[] = tr(" ? is required", array($r));
       }
     }     
     if( $log_action == 'LABOR' ) {
       if( !$hours )
-	$errs[] = tr("Hours must be entered if the activity is 'LABOR'");
+      $errs[] = tr("Hours must be entered if the activity is 'LABOR'");
     } else if( !$comments ) {
       $errs[] = tr("No comments were entered.");
     }
@@ -35,17 +35,17 @@
     if( !$errs ) {
       $res = $zen->log_ticket($id, $login_id, $log_action, $hours, $comments);
       if( $res ) {
-	add_system_messages(tr("Activity has been logged."));
-	$setmode = "log";
-	include("../ticket.php");
-	exit;
-	//header("Location:$rootUrl/ticket.php?id=$id&setmode=log");
+        add_system_messages(tr("Activity has been logged."));
+        $setmode = "log";
+        include("../ticket.php");
+        exit;
+        //header("Location:$rootUrl/ticket.php?id=$id&setmode=log");
       } else {
-	$errs[] = tr("System error: Activity could not be logged.").$zen->db_error;
+        $errs[] = tr("System error: Activity could not be logged.").$zen->db_error;
       }
     }
     if( $errs )
-      add_system_messages( $errs, 'Error' );     
+    add_system_messages( $errs, 'Error' );     
   }
   
   include("$libDir/nav.php");
@@ -64,5 +64,5 @@
   }
   
   include("$libDir/footer.php");
-
-}?>
+  
+}?>
\ No newline at end of file
diff -Naur zentrack_2-5-5-1/www/addProjectSubmit.php zentrack_2-5-5-2/www/addProjectSubmit.php
--- zentrack_2-5-5-1/www/addProjectSubmit.php	2005-05-01 23:59:39.000000000 -0700
+++ zentrack_2-5-5-2/www/addProjectSubmit.php	2005-05-15 16:45:03.000000000 -0700
@@ -22,26 +22,9 @@
   $view = 'project_create';
   
   // initiate default values
-  $otime = time();  // set time ticket opened
-  if( $deadline ) {
-    $deadline = $zen->dateParse($deadline);
-  }
-  if( $start_date ) {
-    $start_date = $zen->dateParse($start_date);
-  }
-  
-  $type_id = $zen->projectTypeID();
-  $description = nl2br($zen->ffv($description));
-  
+  $otime = date('Y-m-d H:i:s');  // set time ticket opened
   include("$libDir/validateFields.php");
-  
-  $zen->cleanInput($fields);
-  // check for required fields
-  foreach($required as $r) {
-    if( !strlen($$r) ) {
-      $errs[] = tr("required field missing:") . " " . ucfirst($r);
-    }
-  }
+
   if( !$errs ) {
     // create an array of existing fields
     foreach(array_keys($fields) as $f) {
@@ -49,6 +32,7 @@
         $params["$f"] = $$f;
       }
     }
+    $params['type_id'] = $zen->projectTypeID();
     $params["creator_id"] = $login_id;
     // add the ticket to db
     $id = $zen->add_ticket($params);
diff -Naur zentrack_2-5-5-1/www/addSubmit.php zentrack_2-5-5-2/www/addSubmit.php
--- zentrack_2-5-5-1/www/addSubmit.php	2005-05-08 14:15:03.000000000 -0700
+++ zentrack_2-5-5-2/www/addSubmit.php	2005-05-15 16:45:03.000000000 -0700
@@ -24,7 +24,7 @@
   $expand_tickets = 1;
   
   // initiate default values
-  $otime = date('Y-m-d h:i:s');  // set time ticket opened
+  $otime = date('Y-m-d H:i:s');  // set time ticket opened
 
   include("$libDir/validateFields.php");
   
diff -Naur zentrack_2-5-5-1/www/admin/fieldMap.php zentrack_2-5-5-2/www/admin/fieldMap.php
--- zentrack_2-5-5-1/www/admin/fieldMap.php	2005-05-04 14:39:36.000000000 -0700
+++ zentrack_2-5-5-2/www/admin/fieldMap.php	2005-05-15 16:45:03.000000000 -0700
@@ -2,26 +2,24 @@
 
   /*
   **  EDIT FIELD MAP
-  **  
+  **
   **  Edit/create/delete/sort entries in the field map
   **
   */
-  
+
   include("admin_header.php");
   $page_title = tr("Edit Field Map");
   include("$libDir/nav.php");
   function fmGetSet( $name, $key, $type = '' ) {
     // don't make updates on rows that don't exist (we may skip some fields occasionally
-    if( !array_key_exists($name, $_POST) ) {
-      return;
-    }
+    if( !array_key_exists($name, $_POST) ) { return; }
     // get the updates array
     global $updates;
     global $zen;
     // process the type
     $t = substr($type,0,1);
-    // handle fields which don't get posted (booleans which are not checked)
-    if( !array_key_exists($key, $_POST[$name]) ) { 
+    // handle fields which do not get posted (booleans which are not checked)
+    if( !array_key_exists($key, $_POST[$name]) ) {
       $updates[$name][$key] = $t == 'b'? 0 : null;
       return;
     }
@@ -41,11 +39,11 @@
     }
     $updates[$name][$key] = $v;
   }
-  
+
   if( !$view || !in_array($view, array_keys($GLOBALS['zt_field_dependencies']['views'])) ) {
     $view = 'ticket_create';
   }
-  
+
   if( $TODO == 'save' && $zen->demo_mode == "on" ) {
     $errs[] = "Your operation was completed, but you cannot edit the field map on a demo site.";
   }
@@ -58,21 +56,21 @@
     $allFields = $map->getFieldMap();
     $updates = $allFields[$view];
     unset($allFields);
-    //Zen::printArray($_POST, 'POST_VALS');
-    
+    //Zen::printArray($_POST, 'POST_VALS'); //debug
+
     $fields = $map->listFieldsForView($view);
     foreach($fields as $f) {
       // collect field properties
       $field = $map->getFieldFromMap($view, $f);
       $tprops = getFmTypeProps($field['field_type']);
-      
+
       // remove deleted sections
       if( $field['field_type'] == 'section' && !isset($_POST[$f]) ) {
         $deletes[$f] = $field['field_map_id'];
       }
       // skip rows that aren't sent
       else if( !isset($_POST[$f]) ) { continue; }
-      
+
       // field_label
       fmGetSet($f, 'field_label', 'text');
       // is_visible
@@ -80,7 +78,7 @@
 
       // num_cols
       fmGetSet($f, 'num_cols', 'int');
-      
+
       // num_rows
       if( !$tprops['multiple'] && !$vprops['multiple'] ) {
         $updates[$f]['num_rows'] = 1;
@@ -89,7 +87,7 @@
 
       // the rest does not apply to sections
       if( $field['field_type'] == 'section' ) { continue; }
-      
+
       // properties for this field
       $fprops = getFmFieldProps($view, ZenFieldMap::fieldName($f));
 
@@ -106,19 +104,18 @@
       // otherwise we post it normally
       else if( array_key_exists('field_type', $_POST[$f]) )
         { fmGetSet($f, 'field_type'); }
-      
+
       // is_required
-      if( $vprops['view_only'] ) { $updates[$f]['is_required'] = false; }
-      else if( $view == 'search_form' ) { $updates[$f]['is_required'] = false; }
+      if( $vprops['view_only'] || $view == 'search_form' ) { $updates[$f]['is_required'] = false; }
       else if( $fprops['always_required'] ) { $updates[$f]['is_required'] = true; }
       else { fmGetSet($f, 'is_required', 'boolean'); }
-      
+
       // default_val has some special considerations
       if( !$vprops['view_only'] ) {
         if( $fprops['default'] ) { fmGetSet($f, 'default_val'); }
       }
     }
-    
+
     foreach($_POST as $k=>$v) {
       if( strpos($k, 'section') === 0 && $vprops['sections'] && !array_key_exists($k,$updates) ) {
         // skip section0, it's a template
@@ -154,7 +151,7 @@
       $res = $map->updateFieldMap($view, $updates, $inserts, $deletes);
       print "<p><b>$res[0] of $res[1] updates were saved</b></p>\n";
     }
-    
+
     //cfalonso changed the following line with the next 3 because it didn't work properly:
     //$fields = $updates;
     $fields = $map->getFieldMap($view);
@@ -162,16 +159,16 @@
     //Zen::printArray($updates,'UPDATES');
     //Zen::printArray($inserts,'INSERTS');
     //Zen::printArray($deletes,'DELETES');
-    
+
   }
   else {
     $fields = $map->getFieldMap($view);
   }
-  
+
   $zen->printErrors($errs);
 
   include("$templateDir/fieldMapForm.php");
-  
+
   include("$libDir/footer.php");
 
 
diff -Naur zentrack_2-5-5-1/www/misc/autologin.php zentrack_2-5-5-2/www/misc/autologin.php
--- zentrack_2-5-5-1/www/misc/autologin.php	1969-12-31 16:00:00.000000000 -0800
+++ zentrack_2-5-5-2/www/misc/autologin.php	2005-05-12 07:05:50.000000000 -0700
@@ -0,0 +1,37 @@
+<?{
+  
+  /*
+  **  STORE PASSWORD IN COOKIE
+  */
+  
+  include_once("../header.php");
+  
+  $page_title = tr("Toggle Auto-Login Feature");
+  $expand_options = 1;
+  
+  $on = array_key_exists('setauto', $_GET) && $_GET['setauto'] == 'on';
+  
+  include("$libDir/nav.php");
+  
+  print "<p><span class='error'>";
+  
+  if( $zen->settings['allow_pwd_save'] != 'on' ) {
+    print tr("This feature has been disabled by the administrator.")."\n";
+  }
+  else {
+    $res = $zen->update_pref($login_id, 'autologin', $on? 1 : 0);
+    
+    if( $res ) {
+      print $on? tr("Your auto-login feature has been turned on.  It will become active after your next login attempt.") : 
+                 tr("Your auto-login feature has been turned off.");
+    }
+    else {
+      print tr("Unable to update your user preferences due to a system error.");
+    }
+  }
+  
+  print "</span></p>\n";
+
+  include("$templateDir/optionsMenu.php");
+  include("$libDir/footer.php");
+}?>
\ No newline at end of file
diff -Naur zentrack_2-5-5-1/www/newProject.php zentrack_2-5-5-2/www/newProject.php
--- zentrack_2-5-5-1/www/newProject.php	2005-03-11 07:37:39.000000000 -0800
+++ zentrack_2-5-5-2/www/newProject.php	2005-05-12 07:05:50.000000000 -0700
@@ -11,7 +11,7 @@
 
   $page_title = tr("Create a New Project");
   $expand_projects = 1;
-  $onLoad[] = "behavior_js.php?formset=newProjectForm&mode=create";
+  $onLoad[] = "behavior_js.php?formset=ticketForm&mode=create";
 
   include("$libDir/nav.php");
 
